Flaky Cypress Tests: Outdated IP API Mock Causes Connection Info Test Failures

# Flaky Cypress Tests: Outdated IP API Mock Causes Connection Info Test Failures

## Problem

The Cypress test suite intermittently fails on tests that depend on `connectionInfo` (specifically `connectionInfo.isKnownVpn`). The test "Multiplayer Normal Paths Omnibus > walks properly" fails when trying to verify that conditional displays based on connection info are rendered correctly.

### Recent Failures

- **PR #1143** ([run #21615164505](https://github.com/Watts-Lab/deliberation-empirica/actions/runs/21615164505)): Failed on TestDisplay32 validation
  ```
  AssertionError: Timed out retrying after 4000ms: Expected to find content: 'TestDisplay32'
  within the element: <button...> but never did.
  ```
  - TestDisplay32 requires `connectionInfo.isKnownVpn === false` to render
  - Test failed all 3 retry attempts (Attempt 1, 2, and 3)

- **PR #1140** ([run #21600509535](https://github.com/Watts-Lab/deliberation-empirica/actions/runs/21600509535)): Failed on TestDisplay02 validation with same pattern
  ```
  AssertionError: Timed out retrying after 4000ms: Expected to find content: 'TestDisplay02'
  within the element: <button...> but never did.
  ```

Both PRs had passing lints and other tests, indicating the failures are environmental/flaky rather than code-related.

## Root Cause

The production code and Cypress test mocks are out of sync:

**Production Code** (client/src/components/hooks.js:16-19):
```javascript
async function loadData() {
  try {
    const ipwhois = await axios.get("https://ipwho.is");
    if (ipwhois.status !== 200) {
```

**Cypress Mock** (cypress/support/commands.js:365-382):
```javascript
Cypress.Commands.add("interceptIpApis", () => {
  cy.intercept("http://ip-api.com/json/", {  // ❌ Wrong endpoint!
    statusCode: 200,
    body: {
      status: "success",
      countryCode: "US",
      timezone: "America/New_York",
      query: "1.1.1.1",
    },
  }).as("ip-api");
```

### Timeline

The IP API endpoint was changed from `ip-api.com` to `ipwho.is` in commit [149dd0ec](https://github.com/Watts-Lab/deliberation-empirica/commit/149dd0ec) (PR #811 - "`reference` syntax"), but the Cypress intercept was never updated.

### Impact

When tests call `cy.interceptIpApis()` (used in multiple test files):
1. The intercept for `ip-api.com` is set up but never matched
2. The code makes a real HTTP request to `ipwho.is`
3. The real API call can:
   - Timeout (causing delays)
   - Fail (leaving `connectionInfo.isKnownVpn` as `undefined`)
   - Return unexpected data
4. Conditional renders that depend on `connectionInfo.isKnownVpn === false` don't display
5. Tests fail when looking for elements like TestDisplay32

## Expected Fix

Update `cypress/support/commands.js` to intercept the correct endpoints:

```javascript
Cypress.Commands.add("interceptIpApis", () => {
  // Update to match current production endpoint
  cy.intercept("https://ipwho.is*", {
    statusCode: 200,
    body: {
      success: true,
      ip: "1.1.1.1",
      type: "IPv4",
      continent: "North America",
      continent_code: "NA",
      country: "United States",
      country_code: "US",
      timezone: {
        id: "America/New_York",
        offset: -18000,
      },
      security: {
        vpn: false,  // This sets isKnownVpn to false
        proxy: false,
        tor: false,
      }
    },
  }).as("ipwho-api");

  // Keep VPN list intercept as-is
  cy.intercept(
    "https://raw.githubusercontent.com/X4BNet/lists_vpn/main/output/vpn/ipv4.txt",
    {
      statusCode: 200,
      body: "1.1.1.1",
    }
  ).as("vpn-list");
});
```

**Note**: The ipwho.is API response format differs from ip-api.com. Check the [ipwho.is documentation](https://ipwho.is/) for the correct response structure.

## Additional Context

### Affected Test Files

The `cy.interceptIpApis()` command is used in:
- cypress/e2e/01_Normal_Paths_Omnibus.js:85
- cypress/e2e/02_Batch_Canceled.js:92
- cypress/e2e/07_Returning_Player.js:61
- (potentially others)

### Connection Info Usage

The `useConnectionInfo()` hook (client/src/components/hooks.js) provides:
- `country` - used for consent form localization
- `timezone` / `timezoneOffset` - used for scheduling
- `isKnownVpn` - used in conditional rendering logic

### Treatment File References

Connection info is used in conditional renders throughout the treatment files, e.g.:
```yaml
- type: prompt
  file: projects/example/testDisplay32.md
  conditions:
    - reference: connectionInfo.isKnownVpn
      comparator: equals
      value: false
```

## Acceptance Criteria

- [ ] Update `cy.interceptIpApis()` to mock `https://ipwho.is` endpoint
- [ ] Verify mock response structure matches ipwho.is API format
- [ ] Run full Cypress test suite locally to verify fix
- [ ] Verify TestDisplay32 and similar connection-dependent displays render in tests
- [ ] Document any API response format changes if needed

--- a/cypress/support/commands.js
+++ b/cypress/support/commands.js
@@ -363,20 +363,42 @@ Cypress.Commands.add(
 );
 
 Cypress.Commands.add("interceptIpApis", () => {
-  cy.intercept("http://ip-api.com/json/", {
+  // Updated to match ipwho.is API (changed from ip-api.com in PR #811)
+  cy.intercept("https://ipwho.is*", {
     statusCode: 200,
     body: {
-      status: "success",
-      countryCode: "US",
-      timezone: "America/New_York",
-      query: "1.1.1.1",
+      success: true,
+      ip: "8.8.8.8", // Use an IP that's NOT in the VPN list so isKnownVpn will be false
+      type: "IPv4",
+      continent: "North America",
+      continent_code: "NA",
+      country: "United States",
+      country_code: "US",
+      region: "New York",
+      region_code: "NY",
+      city: "New York",
+      latitude: 40.7128,
+      longitude: -74.0060,
+      timezone: {
+        id: "America/New_York",
+        abbr: "EST",
+        is_dst: false,
+        offset: -18000,
+        utc: "-05:00",
+      },
+      security: {
+        vpn: false,
+        proxy: false,
+        tor: false,
+        relay: false,
+      },
     },
-  }).as("ip-api");
+  }).as("ipwho-api");
   cy.intercept(
     "https://raw.githubusercontent.com/X4BNet/lists_vpn/main/output/vpn/ipv4.txt",
     {
       statusCode: 200,
-      body: "1.1.1.1",
+      body: "1.1.1.1", // VPN list contains 1.1.1.1, but IP is 8.8.8.8, so isKnownVpn will be false
     }
   ).as("vpn-list");
 });

